File: SE324_Parallel_Algorithms/PA_LAB/Lab_1.py
#To show the execution speed of sorting algos 
import random
import time

def GenerateList(r,n):
    
    arr = random.sample(range(1,r),n)
    return arr

def SelectSort(arr):
    
    for i in range(0,len(arr)):
        min_in = i
        for j in range(i+1,len(arr)):
            if(arr[j] < arr[min_in]):
                min_in = j
        arr[min_in], arr[i] = arr[i], arr[min_in]
    

def Part(arr, low, high):
    i = low - 1
    piv = arr[high]
    for j in range(low, high):
        if (arr[j] <= piv):
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i+1],arr[high] = arr[high], arr[i+1]
    return i+1

def QuickSort(arr, low, high):
    if(low < high):
        pi = Part(arr,low,high)
        QuickSort(arr,low,pi-1)
        QuickSort(arr,pi+1,high)


def main():
    
    r = 10000000000
    n = int(input("Enter the number of elements : "))
    inc = int(input("Enter the increments in elements : "))

    for i in range (0,inc):
        arr = GenerateList(r,n*(10**i))
        start = time.time()
        SelectSort(arr)
        end = time.time()
        print("sorted array of {0} elements in {1} using select sort".format(n*(10**i),round(end-start,3)))
        arr = GenerateList(r,n*(10**i))
        start = time.time()
        QuickSort(arr,0,len(arr)-1)
        end = time.time()
        print("sorted array of {0} elements in {1} using quick sort".format(n*(10**i),round(end-start,3)))

if __name__ == '__main__':
    main()

File: SE324_Parallel_Algorithms/PA_LAB/Lab_1.cpp
#include<bits/stdc++.h>
using namespace std;

void printvec(vector<int> v)
{
    for(int i = 0 ; i < v.size() ; i++)
        printf("%d ",v[i]);
    printf("\n");
}

void SelectSort(vector<int> &arr)
{
    int len = arr.size();
    int min_in;
    for(int i = 0; i < len ; i++)
    {
        min_in = i;
        for(int j = i+1 ; j < len ; j++)
        {
            if(arr[j] < arr[min_in])
                min_in = j;
        }
        swap(arr[i],arr[min_in]);
    }
}

vector<int> GenerateList(int n)
{
    vector<int> arr;
    for(int i = 0; i < n ; i++)
        arr.push_back(i+1);
    random_shuffle(arr.begin(),arr.end());
    return arr;
}

int main()
{
    int r,n,inc;
    cout<<"Enter the number of elements : ";
    cin>>n;
    cout<<"Enter the increment : ";
    cin>>inc;
    vector<int> arr = GenerateList(n);
    // printvec(arr);
    time_t start,end;
    time(&start);
    SelectSort(arr);
    time(&end);
    double diff = difftime(end,start);
    // printvec(arr);
    cout<<"Sorted in : "<<diff<<"s "<<endl;
    return 0;
}